lib/id3c/cli/command/etl/manifest.py

```diff
@@ -204,17 +224,24 @@ def upsert_sample(db: DatabaseSession,
         sample = samples[0]
 
         LOG.info(f"Updating existing sample {sample.id}")
-        sample = db.fetch_row("""
-            update warehouse.sample
-               set identifier = %(identifier)s,
-                   collection_identifier = %(collection_identifier)s,
-                   collected = date_or_null(%(collection_date)s),
-                   details = coalesce(details, '{}') || %(additional_details)s
 
-             where sample_id = %(sample_id)s
+        # Create the update statement dynamically depending on the flags that indicate whether
+        # the record contained the respective data.
 
-            returning sample_id as id, identifier, collection_identifier, encounter_id
-            """,
+        update_statement = "update warehouse.sample set details = coalesce(details, '{}') || %(additional_details)s "
+
+        # Only update the identifier or collection_identifier columns if the source data provides a sample barcode
+        if record_has_sample:
+            update_statement += ", identifier = %(identifier)s "
+            if record_has_collection:
+                update_statement += ", collection_identifier = %(collection_identifier)s "
+
+        if record_has_collected_date:
+            update_statement += ", collected = date_or_null(%(collected_date)s) "
+
+        update_statement += "where sample_id = %(sample_id)s returning sample_id as id, identifier, collection_identifier, encounter_id "
```

Comment:
I know this is written like this to avoid repeating code, but this seems more convoluted and confusing than just writing out separate SQL queries

Reply:
I don't think it's especially convoluted to create SQL command text dynamically. If I were to hardcode the SQL for the various scenarios I'd be repeating a good chunk of code 5 times.
sample + no collection + no collected date
sample + collection + no collected date
sample + collection + collected date
no sample + no collected date
no sample + collected date

Reply:
Instead of doing string concatenation, could we do a string substitution, coupling @joverlee521 and @jccraft-uw 's ideas?

e.g.
```
record_has_sample = ", identifier = %(identifier)s " if "sample" in manifest_record.document else ""
...
query_string = f"""
   update warehouse.sample set details = coalesce(details, '{}') || %(additional_details)s 
   {record_has_sample}
   ...
"""
```
and so on for the other boolean flags

Reply:
I've gone with coalescing the identifier and collection_identifier columns. Thomas recently added in the coalesce in the update for collected. This obviates those flag parameters and the SQL command string building.